### PR TITLE
Refactor CLI handling

### PR DESCRIPTION
**Pull request by bkchr**
This pr switches the CLI handling completely to `StructOpt`.

This also introduces a breaking change for the cli interface, to start a node you will now need to type:
`substrate run` instead of just typing `substrate`.

@shawntabrizi this requires some documentation updates.

### CODE REVIEW

**Comment by bkchr** [comment]
Node template update: https://github.com/paritytech/substrate-node-template/pull/20

**Comment by bkchr** [comment]
Polkadot update: https://github.com/paritytech/polkadot/pull/70

**Comment by gnunicorn** [comment]
As discussed offline, let's change this to allow for two extensions (by the `node`) which were possible before. Node should be able to easily provide:
 - [ ] its own sub-commands
 - [ ] extend the parameters provided to the `run` command

**Comment by xlc** [comment]
Instead of subcommand, it is possible to make them a separate cli tool?
Reasons:
1. It will be good to avoid breaking change on the most commonly used feature (i.e. run the node)
2. It make maintain substrate-based node cli tool easier. That it doesn't need to worry about all other non-core functionality (i.e. only need to able to run node) and the utility cli tool can be used to do operations to chain db (e.g. purge, revert, import etc)

**Comment by bkchr** [comment]
@gnunicorn done both :) please review :)

@xlc I don't understand what you mean by `separate cli tool`? You mean extra executable just for `export-blocks` or `import-blocks`?

File: core/cli/src/lib.rs [hunk 1]
```diff
@@ -219,160 +313,94 @@ where
...
-	config.database_path = db_path(&base_path, config.chain_spec.id()).to_string_lossy().into();
-	config.database_cache_size = match matches.value_of("database_cache_size") {
-		Some(s) => Some(s.parse().map_err(|_| "Invalid Database Cache size specified")?),
-		_=> None
-	};
-	config.pruning = match matches.value_of("pruning") {
-		Some("archive") => PruningMode::ArchiveAll,
+	config.database_path =
+		db_path(&base_path, config.chain_spec.id()).to_string_lossy().into();
+	config.database_cache_size = config.database_cache_size;
```
Comment (cmichi) [inline core/cli/src/lib.rs]: This line has no effect, was the assignment maybe intended to be something else?

**Comment by bkchr** [comment]
I can look into `StructOpt` and see if I can fix this.

**Comment by rphmeier** [comment]
Marked as inprogress because the thread indicates that this is blocked due to some pending structopt work.

**Comment by gavofyork** [comment]
@bkchr is this something that could be farmed out to a new starter?

**Comment by bkchr** [comment]
I finished all my stuff, so this would be my next task from tomorrow.

**Comment by bkchr** [comment]
Anyone wants to give it a review?

File: README.adoc [hunk 2]
```diff
@@ -148,21 +148,21 @@ Now with this new chainspec file, you can build a "raw" chain definition for you
...
-substrate --chain ~/mychain.json
+substrate run --chain ~/mychain.json
 ----
 
 It won't do much until you start producing blocks though, so to do that you'll need to use the `--validator` option together with passing the seed for the account(s) that is configured to be the initial authorities:
 
 [source, shell]
 ----
-substrate --chain ~/mychain.json --validator --key ...
+substrate run --chain ~/mychain.json --validator --key ...
```
Comment (gavofyork) [inline README.adoc]: incorrect, no?

File: README.adoc [hunk 3]
```diff
@@ -148,21 +148,21 @@ Now with this new chainspec file, you can build a "raw" chain definition for you
...
-substrate --chain ~/chainspec.json build-spec --raw > ~/mychain.json
+substrate build-spec --chain ~/chainspec.json --raw > ~/mychain.json
 ----
 
 This can be fed into Substrate:
 
 [source, shell]
 ----
-substrate --chain ~/mychain.json
+substrate run --chain ~/mychain.json
```
Comment (gavofyork) [inline README.adoc]: incorrect?

File: README.adoc [hunk 4]
```diff
@@ -124,7 +124,7 @@ To get going as fast as possible, there is a simple script that installs all req
 curl https://getsubstrate.io -sSf | bash
 ----
 
-You can start a local Substrate development chain with running `substrate --dev`.
+You can start a local Substrate development chain with running `substrate run --dev`.
```
Comment (gavofyork) [inline README.adoc]: incorrect?

**Comment by gavofyork** [comment]
Cool - just some doc comments that I think are outdated.

**Comment by gavofyork** [comment]
@bkchr need instructions & updates PRs for node-template & polkadot now.

**Comment by bkchr** [comment]
@gavofyork Yeah, already working on Polkadot update anyway.

**Pull request merged**